### ISSUE TITLE
Better cache locality for thread caches

### ISSUE THREAD

**Issue by toddlipcon** [gperftools/gperftools#951]
Currently the thread cache freelist is implemented using a singly-linked list. This means that, when we transfer elements from the free list back onto the central list, we have to walk through O(n) nodes to accumulate the list of objects to return. Often these have fallen out of CPU cache by the time they are returned, so this walking can be somewhat slow. In fact, in some of our workloads, we see tcmalloc::ThreadCache::ReleaseToCentralCache one of the top consumers of CPU.

Additionally, it returns the N most recently used items from the cache rather than N least recently used, which seems like it might be harmful to spatial locality.

I've done a bit of prototyping of the following idea:

For all but the smallest size class, there's space for more than one pointer in the allocated items. Thus, consider making a singly-linked list of nodes, each of which contains an array of some number of pointers:

struct Node {
  Node* next;
  int item_count; // actually can be inlined in the high 16 bits of the 'next' pointer
  Node* items[N]; // actually a VLA based on the size of freelist items
}

Now we get the following pseudocode:
Push(elem):
- if the current head has space for another pointer, add elem to the end of the 'items' array
- otherwise, initialize 'elem' as a new Node with 0 inline items, and link its 'next' pointer to the current head

Pop(elem):
- basically the inverse of the above

BatchPop(elem, N):
- while N > the number of elements in the head node:
-- dequeue all the elements in the head node
-- if necessary, dequeue the head node itself, exposing a new head node

I wrote a simple prototype of this data structure and for 128-byte allocations it can push about 5x faster and batch-dequeue about 3x faster in some dirty benchmarks.

I haven't yet tried hooking this into tcmalloc itself yet, but wanted to get feedback on the approach: would this kind of change be considered?

**Comment by toddlipcon** [comment]
I did a little more hacking on this and hooked the implementation into tcmalloc. So far the results are mixed on malloc_benchmark:

Most of the benchmarks there use a very small working set -- eg allocating a few hundred items, freeing them, and continuing in a loop. The linked list which represents the freelist in current versions of tcmalloc would entirely fit in cache then -- eg with 512 items, we'll assume 512x64b cache line = 32kb worth of cache space needed, which is entirely L1-resident. So, the cache locality benefits of this improvement don't show up, and the extra "branchiness" ends up hurting 20-40%.

There's a pretty big win on the benchmarks with larger working sets, though. Most notably:
Before:
Benchmark: bench_fastpath_stack_simple(8192)                20.102893 nsec
Benchmark: bench_fastpath_stack_simple(8192)                20.103229 nsec
Benchmark: bench_fastpath_stack_simple(8192)                20.168054 nsec
Benchmark: bench_fastpath_rnd_dependent(8192)               37.179026 nsec
Benchmark: bench_fastpath_rnd_dependent(8192)               36.990279 nsec
Benchmark: bench_fastpath_rnd_dependent(8192)               37.101237 nsec

After:
Benchmark: bench_fastpath_stack_simple(8192)                8.174208 nsec
Benchmark: bench_fastpath_stack_simple(8192)                8.146939 nsec
Benchmark: bench_fastpath_stack_simple(8192)                8.099899 nsec
Benchmark: bench_fastpath_rnd_dependent(8192)               19.507631 nsec
Benchmark: bench_fastpath_rnd_dependent(8192)               19.312914 nsec
Benchmark: bench_fastpath_rnd_dependent(8192)               19.426305 nsec

I started doing some benchmarking with larger working set sizes but seeing some strange behavior -- eg the original version actually gets _faster_ with a larger working set, which doesn't make much sense. Perhaps the transfer cache is more effective at that larger size, or some other effect is kicking in. Figured I'd report progress here to see if folks think this is worth pursuing, and whether there is a better (more realistic) workload to play with.

**Comment by alk** [comment]
Thanks a lot. Yes, this is very relevant.

In fact very similar if not 100% same idea have been explored semi-recently at Google. Let me CC relevant people (@ahh, @ckennelly, @dvyukov, although I am not sure github will actually inform them).

So, yes, we're well aware of "pointer chasing" issues with thread caches mode. There is nothing going on for that at google, because per-cpu cache mode doesn't have this problem.

You may have heard about restartable sequences (e.g. https://lwn.net/Articles/697979/) and how Google-internal tcmalloc fork is using it to have per-cpu caches. This is no secret code, but it wasn't shared because it requires special kernel support (and special glibc support and weird build system, etc). The issue for gperftools is that even when Linux gets restartable sequences support, we'll have to support decent per-thread mode for many years. So this is very relevant.

Per-cpu mode actually has different representation of freelists and is much faster in practice. Because we have limited number of cpus, we can afford having just statically sized array of pointers for every size class. Then pushing or popping object into/from freelist doesn't need to touch object itself. But most importantly it lets us push and pop whole batch of object pointers without touching any extra memory.

Google now runs almost 100% of workloads in per-cpu mode which eliminates this issue. So nothing has been done about per-thread free lists other than few experiments.

So Chris actually implemented idea that looks very similar to yours. IMHO the issue with this approach is that it is very branchy and there is not much space in very common smaller size classes to really make it fast. So overhead is relatively high. There might be some apps that still get faster despite overhead due to fewer pointer chasing. AFAIK Chris has seen some app benchmark improved with his patch.

Also I briefly tried what looked like slightly lighter version of this idea. Basically, instead of having one linked list per size class, we have two and we interleave popping from them. It helps with pointer chasing because we let cpu work on popping both lists at a time. I.e. either as part of mass-pop/push in slow path, or in fast-path when malloc follows malloc and cpu can speculate past cache miss in first malloc and actually perform most of second malloc. This option was also making those larger working set benchmarks faster. But fast-path overhead was still somewhat elevated. I never went past few micro-benchmark runs with that code. Maybe we could revisit it.

In past I also toyed with another free list idea that is a bit difficult for me to describe. But it was also too branchy.

Another thing worth mentioning is that google's tcmalloc has very similar implementation of central free list. Since free list only needs refer to object within one tcmalloc page. Dmitry implemented free list like you're describing but using 16 bits "offsets" instead. Which lets us have 4 logical pointers per even 8 bytes size class. People had some other ideas too about central free list. I looked at porting this code as part of gperftools 2.6 work, but because we have pointer chasing it per-thread free lists, this data structure doesn't seem to help. So I left it out. For now at least.

So indeed there are lots of interesting options that are worth exploring. We lack good benchmarks to evaluate those ideas.

That is why I was recently working on fast and accurate malloc tracing facility. I encourage you to take a look at the doc: https://docs.google.com/document/d/1S8T9apA2juEC5sV2UVXUan4e4_029fTmoAEoX7MEB9s and see if you can capture malloc traces from your workloads. It would be super if you could share your traces.

The thinking is with wide range of relevant malloc traces we should be able to evaluate a number of design ideas with more confidence. Which includes options for free lists data structures.

I am thinking there might be some way to have most active small subset of per-thread caches using arrays for smallest free lists. That would be fastest I think. But will depend on whether we can make memory-space overhead low enough.

Or maybe we should look at what jemalloc folks have been experimenting with "user-space" rseq. I.e. here https://github.com/facebookexperimental/Rseq/blob/master/Rseq.md and here: https://www.pvk.ca/Blog/2017/06/05/relaxed-revocable-locks-mutual-exclusion-modulo-preemption/.

**Comment by toddlipcon** [comment]
Hey Aliaksey. Thanks for the detailed response.

Regarding the per-CPU mode and user-space Rseq, I've indeed been following both the upstream kernel discussions and Facebook's work within jemalloc. lnterestingly, the last I heard was that they hadn't gotten the jemalloc changes to give positive perf results on real applications at FB yet, but things may have changed in the last 6 months or so since I last looked into it in any detail. I definitely think it would be great if rseq made it outside Google -- we often have to spend quite a bit of effort to "'balance" allocations across threads, which can get particularly tough with typical producer-consumer relationships between RPC service pools and various workers within server apps.

As for the central freelist using page-relative pointers, that's quite a clever idea and something I didn't consider. It does seem like it would help quite a bit for the small allocation case. Another thing I considered (but didn't implement) was to have the small size class actually steal allocations from a larger size class (eg 64-byte) to build a non-intrusive stack. I was afraid of code complexity there, though, so didn't pursue it yet.

To follow up on the implementation I'd mentioned a couple weeks ago, the unexpected perf results were actually from my not reading the benchmark code carefully. The larger sizes in the benchmark actually "wrapped" back down to being very small working set sizes due to some bit-masking of the benchmark parameters. I did a few tricks in the code to reduce the number of branches and was able to get a decent speedup on large working sets and small hit on small ones. I definitely think the tracing work you did will be very helpful - I'll check that out in the next week or two.

Only concern I had about the tracing is that it may not well represent the cache behavior of real applications. For example, in some of our applications we have cases where we free pointers on a thread that has never actually accessed those pointers. This means that if tcmalloc's free path actually stores a linked list within the freed memory it will cause a cache miss at free time, whereas an implementation that doesn't access the pointer would not. If the trace replay code doesn't capture the memory access behavior of the app, we may not see a perf difference between the two candidate implementations.

Unfortunately fully tracing and simulating the memory behavior of an application may be quite difficult. I wonder whether something like the LLVM efficiency sanitizer could be of use.

I suppose one possibility is a set of coarse flags for the trace replay code such as "--read-memory-before-free" or "--write-memory-before-free", "--read-memory-after-malloc", etc. 

-Todd

**Comment by alk** [comment]
Reproduced "duplex" idea I mentioned above for gperftools. It seems to affect fast-path microbenchmark not too much and does help for pointer chasing. Seems to speed up replay of malloc trace I have of google search backend. Of course, it should help even more once we integrate better central free list code.

Code here: https://github.com/alk/gperftools/tree/wip-duplex-freelist

**Comment by toddlipcon** [comment]
I tried running the 'duplex' patch with a trace replay from Kudu (the same one I sent you off-list) and it seems to regress a bit (+8% cycle count)

Before:
```
total_instructions = 1023410420; rate = 35776085.653663 ops/sec; live threads: 163; ~last reg: 37279
processed total 1025163027 malloc ops (aka instructions)
toal thread switches 25241330
total cpu switches 25008449
total ts updates 25275373
allocated still: 0

 Performance counter stats for './replay':

      23157.069072      task-clock (msec)         #    0.807 CPUs utilized
             6,051      context-switches          #    0.261 K/sec
                 0      cpu-migrations            #    0.000 K/sec
           176,876      page-faults               #    0.008 M/sec
    76,427,832,248      cycles                    #    3.300 GHz                      (62.38%)
    96,104,750,843      instructions              #    1.26  insn per cycle           (74.90%)
    20,345,777,829      branches                  #  878.599 M/sec                    (75.06%)
       210,381,386      branch-misses             #    1.03% of all branches          (75.10%)
    35,910,904,720      L1-dcache-loads           # 1550.753 M/sec                    (75.07%)
     2,635,038,089      L1-dcache-load-misses     #    7.34% of all L1-dcache hits    (75.05%)
       791,858,500      LLC-loads                 #   34.195 M/sec                    (49.87%)
         3,324,573      LLC-load-misses           #    0.42% of all LL-cache hits     (49.85%)

      28.702480205 seconds time elapsed
```

After:
```
total_instructions = 1023410420; rate = 35795702.845527 ops/sec; live threads: 163; ~last reg: 37279
processed total 1025163027 malloc ops (aka instructions)
toal thread switches 25241330
total cpu switches 25008449
total ts updates 25275373
allocated still: 0

 Performance counter stats for './replay':

      24976.676486      task-clock (msec)         #    0.871 CPUs utilized
             5,914      context-switches          #    0.237 K/sec
                 1      cpu-migrations            #    0.000 K/sec
           176,891      page-faults               #    0.007 M/sec
    82,640,271,176      cycles                    #    3.309 GHz                      (62.52%)
   104,273,614,936      instructions              #    1.26  insn per cycle           (74.99%)
    20,850,050,774      branches                  #  834.781 M/sec                    (75.07%)
       219,927,104      branch-misses             #    1.05% of all branches          (75.02%)
    37,907,062,018      L1-dcache-loads           # 1517.698 M/sec                    (75.02%)
     2,788,795,133      L1-dcache-load-misses     #    7.36% of all L1-dcache hits    (75.01%)
       837,572,128      LLC-loads                 #   33.534 M/sec                    (49.91%)
         3,428,843      LLC-load-misses           #    0.41% of all LL-cache hits     (49.97%)

      28.686179818 seconds time elapsed
```

**Comment by obones** [comment]
Do restartable sequences exist in the Windows world?
I'm asking because we have very good results under Windows 7 and upper and are keeping an alert eye on upcoming changes.

**Comment by alk** [comment]
rseq isn't available on windows. Well, even on Linux is is only about to get available with upcoming release. It won't get to distro kernels for awhile. So we'll keep supporting thread caches mode. And if I will find a way to make thread caches perform just as good, I'll be actually killing per-cpu mode.

**Comment by obones** [comment]
Thanks for the details, it's good to know there are ideas floating around
